model/sighandler/sighandler_timestamp.go

```diff
@@ -0,0 +1,206 @@
+		if certObj.Len() == 0 {
+			return nil, errors.New("no signature certificates found")
+		}
+		for _, obj := range certObj.Elements() {
+			certStr, ok := core.GetString(obj)
+			if !ok {
+				return nil, fmt.Errorf("invalid certificate object type in signature certificate chain: %T", obj)
+			}
+			certData = append(certData, certStr.Bytes()...)
+		}
+	default:
+		return nil, fmt.Errorf("invalid signature certificate object type: %T", certObj)
+	}
+
+	certs, err := x509.ParseCertificates(certData)
+	if err != nil {
+		return nil, err
+	}
+
+	return certs[0], nil
+}
+
+// NewDigest creates a new digest.
+func (a *docTimeStamp) NewDigest(sig *model.PdfSignature) (model.Hasher, error) {
+	return bytes.NewBuffer(nil), nil
+}
+
+type timestampInfo struct {
+	Version        int
+	Policy         asn1.RawValue
+	MessageImprint struct {
+		HashAlgorithm pkix.AlgorithmIdentifier
+		HashedMessage []byte
+	}
+}
+
+func getHashForOID(oid asn1.ObjectIdentifier) (crypto.Hash, error) {
+	switch {
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA1), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA1),
+		oid.Equal(pkcs7.OIDDigestAlgorithmDSA), oid.Equal(pkcs7.OIDDigestAlgorithmDSASHA1),
+		oid.Equal(pkcs7.OIDEncryptionAlgorithmRSA):
+		return crypto.SHA1, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA256), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA256):
+		return crypto.SHA256, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA384), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA384):
+		return crypto.SHA384, nil
+	case oid.Equal(pkcs7.OIDDigestAlgorithmSHA512), oid.Equal(pkcs7.OIDDigestAlgorithmECDSASHA512):
+		return crypto.SHA512, nil
+	}
+	return crypto.Hash(0), pkcs7.ErrUnsupportedAlgorithm
+}
+
+// Validate validates PdfSignature.
+func (a *docTimeStamp) Validate(sig *model.PdfSignature, digest model.Hasher) (model.SignatureValidationResult, error) {
+	signed := sig.Contents.Bytes()
+	p7, err := pkcs7.Parse(signed)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	if err = p7.Verify(); err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	var tsInfo timestampInfo
+
+	_, err = asn1.Unmarshal(p7.Content, &tsInfo)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+
+	hAlg, err := getHashForOID(tsInfo.MessageImprint.HashAlgorithm.Algorithm)
+	if err != nil {
+		return model.SignatureValidationResult{}, err
+	}
+	h := hAlg.New()
+	buffer := digest.(*bytes.Buffer)
+
+	h.Write(buffer.Bytes())
+	sm := h.Sum(nil)
+	bytes.Equal(sm, tsInfo.MessageImprint.HashedMessage)
+
+	return model.SignatureValidationResult{IsSigned: true, IsVerified: bytes.Equal(sm, tsInfo.MessageImprint.HashedMessage)}, nil
+}
+
+// Sign sets the Contents fields for the PdfSignature.
+func (a *docTimeStamp) Sign(sig *model.PdfSignature, digest model.Hasher) error {
+	if a.emptySignatureLen <= 0 {
+		sig.Contents = core.MakeHexString(string(make([]byte, 8192)))
+		return nil
+	}
+
+	buffer := digest.(*bytes.Buffer)
+
+	h := crypto.SHA512.New()
```

Comment:
I think it would be better to use `a.hashAlgorithm` here and when creating the `timestamp.Request`. Maybe we can add an option to configure the hashing algorithm when creating the signature handler, as the validation already seems to handle multiple hashers. 
This could be added either through an additional parameter to the `NewDocTimeStamp`, or by implementing an additional handler creation function which has a parameter for configuring the hashing algorithm.